net_util/src/queue_pair.rs

```diff
@@ -128,63 +107,64 @@ impl RxVirtio {
     pub fn process_desc_chain(
         &mut self,
         mem: &GuestMemoryMmap,
-        mut next_desc: Option<DescriptorChain>,
+        tap: &mut Tap,
         queue: &mut Queue,
-    ) -> bool {
-        let head_index = next_desc.as_ref().unwrap().index;
-        let mut write_count = 0;
-
-        // Copy from frame into buffer, which may span multiple descriptors.
-        loop {
-            match next_desc {
-                Some(desc) => {
-                    if !desc.is_write_only() {
-                        break;
-                    }
-                    let limit = cmp::min(write_count + desc.len as usize, self.bytes_read);
-                    let source_slice = &self.frame_buf[write_count..limit];
-                    let write_result = mem.write_slice(source_slice, desc.addr);
+    ) -> Result<bool, NetQueuePairError> {
+        let mut exhausted_descs = true;
+        while let Some(avail_desc) = queue.iter(&mem).next() {
+            let head_index = avail_desc.index;
+            let mut next_desc = Some(avail_desc);
 
-                    match write_result {
-                        Ok(_) => {
-                            write_count = limit;
-                        }
-                        Err(e) => {
-                            error!("Failed to write slice: {:?}", e);
-                            break;
-                        }
+            let mut iovecs = Vec::new();
+            while let Some(desc) = next_desc {
+                if desc.is_write_only() {
+                    let buf = mem
+                        .get_slice(desc.addr, desc.len as usize)
+                        .map_err(NetQueuePairError::GuestMemory)?
+                        .as_ptr();
+                    let iovec = libc::iovec {
+                        iov_base: buf as *mut libc::c_void,
+                        iov_len: desc.len as libc::size_t,
                     };
-
-                    if write_count >= self.bytes_read {
-                        break;
-                    }
-                    next_desc = desc.next_descriptor();
-                }
-                None => {
-                    warn!("Receiving buffer is too small to hold frame of current size");
-                    break;
+                    iovecs.push(iovec);
                 }
+                next_desc = desc.next_descriptor();
             }
-        }
 
-        self.counter_bytes += Wrapping((write_count - vnet_hdr_len()) as u64);
-        self.counter_frames += Wrapping(1);
+            let len = if !iovecs.is_empty() {
+                let result = unsafe {
+                    libc::readv(
+                        tap.as_raw_fd() as libc::c_int,
+                        iovecs.as_ptr() as *const libc::iovec,
+                        iovecs.len() as libc::c_int,
+                    )
```

Comment:
One of the reasons why there is an intermediate buffer is because when you ask to read from a TAP device you must read the whole frame in one go; if you read only a fraction then the remainder is "lost". I don't see how you handle that situation.

Reply:
https://github.com/cloud-hypervisor/cloud-hypervisor/issues/1265#issuecomment-638781834

Reply:
Or is the kernel providing a descriptor chain of <header size> <MTU size> <header size> <MTU size> ... ?

Reply:
The TAP always provide whole frames, the problem was coming from the way we were reading/writing each time into a descriptor, which was not matching the size of the frame.
By using vectorized read/write, we solve the problem as the whole frame can be read/written in one go.

Reply:
Okay, cool. What did you have to fix for the CI to pass?

Reply:
I was missing the seccomp filters.